content/blog/2025-05-13-polyhedral.md

```diff
@@ -0,0 +1,190 @@
+their algorithms are: [IJPP'00][ijpp-00], [PACT'04][pact-04], [TOPLAS'15][toplas-15].
+
+[ijpp-00]: https://link.springer.com/article/10.1023/A:1007554627716
+[pact-04]: https://icps.u-strasbg.fr/~bastoul/research/papers/Bas04-PACT.pdf
+[toplas-15]: https://lirias.kuleuven.be/retrieve/319112
+
+Because of this (and other reasons), [MLIR introduced][mlir] a _simplified_
+polyhedral representation, which has an explicit representation of iteration
+domains, and the analysis / transformation benefits that affords, but also a
+preferred way of iterating over it.
+
+I copied their ideas to design my own IR, which has iteration domains
+represented as integer sets (using ISL, the Integer Set Library), but also has
+a relatively straightforward interpretation as code:
+
+```
+s ::= for (i ∈ affine domain) { ss }        (single-variable for loop)
+   |  if (affine condition) { ss }          (conditionals)
+   |  instr_1; ...; instr_N                 (basic blocks)
+
+ss ::= s_1; ...; s_N
+```
+
+Here, both `affine domain` and `affine condition` are integer sets; the only
+difference is that `affine domain` introduces a new variable, while `affine
+condition` does not.
+
+A core insight is that we have _static control flow_: there is a subset of
+variables, which I'll call "affine variables", which are only transformed in
+very easy-to-reason-about ways, and moreover, control flow only depends on
+these variables. Of course, not every program has this format, but we still
+want to be able to optimize the programs that do, while leaving those that
+don't.
+
+## To `affine`-ity, and beyond!
+
+We've made the process of emitting code out of our IR easier, with the
+simplified polyhedral representation. But how do we get it there in the first
+place? This is still fairly tricky!
+
+Static control flow is a subset of structured control flow, so it makes sense
+to start with a program with structured control flow. Bril doesn't have
+structured control flow, but we know that it's possible to construct if you
+have reducible control flow. So the first step is to reconstruct the structured
+control flow of the input Bril program. This is the job of the relooper.
+
+Given a Bril function as input, the relooper either:
+
+* raises an exception, if it has irreducible control flow, or
+* returns a Bril AST with structured control flow.
+
+Here's what this structured control flow IR looks like:
+
+```
+c ::= block .l { c1 } c2            (l is in scope for c1 but not c2)
+   |  loop .l { c }                 (l is in scope for c)
+   |  instr_1; ...; instr_N; c
+   |  if b then c else c
+   |  jmp .l
+   |  ret v
+```
+
+A jump to label `.l` is only allowed when `.l` is in scope (this is what makes
+it structured!). Jumping to a `block` label is like a `break`, escaping to the
+end of the block, while jumping to a `loop` label is like a `continue`, going
+back to the loop header for the next iteration.
```

Comment:
This is a clever way to repurpose Bril labels and `jmp` in a structured representation.